packages/frontend/workspace/src/providers/index.ts

```diff
@@ -1,160 +1,114 @@
-    },
-    subscribeStatusChange: indexeddbProvider.subscribeStatusChange,
-    get connected() {
-      return connected;
-    },
-    cleanup: () => {
-      indexeddbProvider.cleanup().catch(console.error);
-    },
-    connect: () => {
-      logger.info('connect indexeddb provider', id);
-      indexeddbProvider.connect();
-    },
-    disconnect: () => {
-      assertExists(indexeddbProvider);
-      logger.info('disconnect indexeddb provider', id);
-      indexeddbProvider.disconnect();
-      connected = false;
-    },
-  };
-};
-
-const indexedDBDownloadOrigin = 'indexeddb-download-provider';
-
-const createIndexedDBDownloadProvider: DocProviderCreator = (
-  id,
-  doc
-): LocalIndexedDBDownloadProvider => {
-  const datasource = createIndexedDBDatasource({});
-  let _resolve: () => void;
-  let _reject: (error: unknown) => void;
-  const promise = new Promise<void>((resolve, reject) => {
-    _resolve = resolve;
-    _reject = reject;
-  });
-
-  return {
-    flavour: 'local-indexeddb',
-    active: true,
-    get whenReady() {
-      return promise;
-    },
-    cleanup: () => {
-      // todo: cleanup data
-    },
-    sync: () => {
-      logger.info('sync indexeddb provider', id);
-      datasource
-        .queryDocState(doc.guid, {
-          stateVector: encodeStateVector(doc),
-        })
-        .then(docState => {
-          if (docState) {
-            Y.applyUpdate(doc, docState.missing, indexedDBDownloadOrigin);
-          }
-          _resolve();
-        })
-        .catch(_reject);
-    },
-  };
-};
-
-export {
-  createAffineSocketIOProvider,
-  createBroadcastChannelProvider,
-  createIndexedDBBackgroundProvider,
-  createIndexedDBDownloadProvider,
-  createSQLiteDBDownloadProvider,
-  createSQLiteProvider,
-  downloadBinaryFromCloud,
-};
+export * from './sync';
 
 export const createLocalProviders = (): DocProviderCreator[] => {
-  const providers = [
-    createIndexedDBBackgroundProvider,
-    createIndexedDBDownloadProvider,
-  ] as DocProviderCreator[];
-
-  if (runtimeConfig.enableBroadcastChannelProvider) {
-    providers.push(createBroadcastChannelProvider);
-  }
-
-  if (environment.isDesktop && runtimeConfig.enableSQLiteProvider) {
-    providers.push(createSQLiteProvider, createSQLiteDBDownloadProvider);
-  }
-
-  return providers;
+  return [
+    (_, doc, { awareness }) => {
+      const engine = new SyncEngine(
+        environment.isDesktop
+          ? createSQLiteStorage(doc.guid)
+          : createIndexedDBStorage(doc.guid),
+        [],
+        doc
+      );
+      engine.start();
+
+      const awarenessProviders = [
+        createBroadcastChannelAwarenessProvider(doc.guid, awareness),
+      ];
+
+      for (const provider of awarenessProviders) {
+        provider.connect();
+      }
+
+      let connected = false;
+
+      return {
+        flavour: '_',
+        passive: true,
+        connect() {
+          connected = true;
+          // TODO: actually connect
```

Comment:
delay the start to here?

Reply:
Maybe in the future? currently we control `connect` and `disconnect` in jotai, but some use workspace outside jotai. I want to refactor this part in the future.

Reply:
if so, would it be better we expose the `SyncEngine` as app level state as well, and let the owner to control the connect/disconnect/start?

Reply:
I just put the connection part into the sync() function and it magically worked. and only connect to the workspace when needed